cql3/restrictions/analyzed_where.cc

```diff
@@ -23,6 +26,12 @@ analyzed_where_clause::analyzed_where_clause(data_dictionary::database db,
     query_restrictions = std::nullopt;
     restrictions = ::make_shared<statement_restrictions>(db, schema, type, where_clause, ctx,
                                                          selects_only_static_columns, for_view, allow_filtering);
+
+    if (!restrictions->uses_secondary_indexing()) {
+        expr::expression prepared_where_clause =
+            expr::prepare_expression(where_clause, db, schema->ks_name(), schema.get(), nullptr);
+        query_restrictions = single_table_query_restrictions::make(prepared_where_clause, schema);
+    }
```

Comment:
I think this conditional prepare is very dangerous. There should be a clear boundary between prepared and unprepared expressions, and it should be passed in in select_statement::prepare(), not deep inside some other functions.

Perhaps this is due to the transition between the old-style restrictions and the new code? If so maybe passing both prepared and unprepared where_clause here (just during the transition) can help. In the end, this will receive just the prepared where clause.

Reply:
Preparing the WHERE clause is currently done inside `statement_restrictions` constructor. The problem is that apart from preparing, it does validation at the same time. Preparing binary_operators and validating them is intertwined, because there are custom error messages for certain types of problems.
Eventually once the grammar is fully relaxed, preparing will equal validation, because any valid expression will describe a valid query. For now statement_restrictions does its thing to validate and prepare the WHERE for its use.

I added a separate `prepare_expression` to keep the new code fully independent from `statement_restrictions`. After constructing `statement_restrictions` we can be sure that the WHERE clause is valid, and `prepare_expression` is enough to get a prepared WHERE. 
The only problem is that the WHERE is prepared two times, I hope this doesn't cause any issues.

Reply:
We should relax the constraint about keeping the same error messages, if it's causing too much trouble. We don't guarantee the error messages will remain the same. So if it helps you, then you can change them as long as they are still informative.

(but maybe changing dtest is even less fun)